src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6373,29 +6380,33 @@ private bool IsTerm()
 
         private TypeSyntax ParseTypeCore(
             bool parentIsParameter,
-            bool isOrAs,
+            bool isPattern,
+            bool rightOfAs,
             bool expectSizes,
             bool isArrayCreation)
         {
-            if (!parentIsParameter && !isOrAs && !expectSizes && !isArrayCreation && this.CurrentToken.Kind == SyntaxKind.RefKeyword)
+            if (!parentIsParameter && !(isPattern || rightOfAs) && !expectSizes && !isArrayCreation && this.CurrentToken.Kind == SyntaxKind.RefKeyword)
             {
                 var refKeyword = this.EatToken();
-                var type = ParseTypeCore1(parentIsParameter, isOrAs, expectSizes, isArrayCreation);
+                var type = ParseTypeCore1(parentIsParameter, isPattern, rightOfAs, expectSizes, isArrayCreation);
                 return this.CheckFeatureAvailability(_syntaxFactory.RefType(refKeyword, type), MessageID.IDS_FeatureRefLocalsReturns);
             }
             else
             {
-                return ParseTypeCore1(parentIsParameter, isOrAs, expectSizes, isArrayCreation);
+                return ParseTypeCore1(parentIsParameter, isPattern, rightOfAs, expectSizes, isArrayCreation);
             }
         }
 
         private TypeSyntax ParseTypeCore1(
             bool parentIsParameter,
-            bool isOrAs,
+            bool rightOfIs,
+            bool rightOfAs,
```

Comment:
It looks like these two parameters are mutually exclusive. Consider adding an assert, or better use one enum parameter instead.